File: dockerfiles/attestation-sidecar/app/internal/blockchain/event_watcher.py
# app/internal/blockchain/event_watcher.py

import json
import time
import os
from typing import Iterable, Optional, Dict, Any
from web3 import Web3
from app.internal.logger import info, warn, error, debug

class EventWatcher:
    def __init__(
        self,
        web3: Web3,
        contract,
        event_abi: dict,
        address: str,
        topics: Optional[list] = None,
        checkpoint_path: str = "event_checkpoint.json",
        confirmations: int = 2,
        batch_size: int = 1000,
        poll_interval: float = 1.0,
    ):
        self.w3 = web3
        self.address = Web3.to_checksum_address(address)
        self.contract = contract
        self.event_abi = event_abi
        self.topics = topics
        self.checkpoint_path = checkpoint_path
        self.confirmations = confirmations
        self.batch_size = batch_size
        self.poll_interval = poll_interval

        cp = self._load_checkpoint()
        self.from_block = cp.get("from_block", self.w3.eth.block_number)
        self.seen_keys = set(cp.get("seen_keys", []))

    def _load_checkpoint(self) -> Dict[str, Any]:
        if os.path.exists(self.checkpoint_path):
            try:
                with open(self.checkpoint_path, "r") as f:
                    data = json.load(f)
                    if isinstance(data, dict):
                        return data
            except (json.JSONDecodeError, OSError):
                pass
        return {}

    def _save_checkpoint(self):
        directory = os.path.dirname(self.checkpoint_path)
        if directory:
            os.makedirs(directory, exist_ok=True)
            
        with open(self.checkpoint_path, "w") as f:
            json.dump(
                {"from_block": self.from_block, "seen_keys": list(self.seen_keys)},
                f,
                indent=2,
            )

    def _latest_safe_block(self) -> int:
        current = self.w3.eth.block_number
        return max(0, current - self.confirmations)

    def _unique_key(self, log) -> str:
        return f"{log['blockHash'].hex()}-{log['logIndex']}"

    def _decode_log(self, raw_log):
        event_name = self.event_abi['name']
        event_cls = getattr(self.contract.events, event_name)
        return event_cls.process_log(raw_log)

    def _fetch_logs_range(self, start_block: int, end_block: int) -> Iterable[dict]:
        # FIX: Use 'fromBlock' and 'toBlock' (CamelCase) for raw RPC compatibility
        filter_params = {
            "fromBlock": start_block,
            "toBlock": end_block,
            "address": self.address,
        }
        if self.topics:
            filter_params["topics"] = self.topics
            
        return self.w3.eth.get_logs(filter_params)

    def run(self, handle_event):
        event_name = self.event_abi.get("name", "UnknownEvent")
        topics_str = str(self.topics) if self.topics else "None"

        info(
            f"EventWatcher start: event={event_name} address={self.address} "
            f"from_block={self.from_block} confirmations={self.confirmations} "
            f"batch_size={self.batch_size} poll_interval={self.poll_interval:.3f}s topics={topics_str}"
        )
        
        while True:
            try:
                safe_to = self._latest_safe_block()
                
                if self.from_block > safe_to:
                    time.sleep(self.poll_interval)
                    continue

                while self.from_block <= safe_to:
                    to_block = min(self.from_block + self.batch_size - 1, safe_to)
                    
                    raw_logs = self._fetch_logs_range(self.from_block, to_block)
                    logs_list = list(raw_logs)
                    logs_list.sort(key=lambda l: (l['blockNumber'], l['logIndex']))

                    for raw in logs_list:
                        key = self._unique_key(raw)
                        if key in self.seen_keys:
                            continue
                        
                        try:
                            evt = self._decode_log(raw)
                            handle_event(evt)
                            self.seen_keys.add(key)
                        except Exception as e:
                            error(f"Error decoding/handling log {key}: {e}")

                    self.seen_keys.clear()
                    self.from_block = to_block + 1
                    self._save_checkpoint()

                time.sleep(self.poll_interval)

            except Exception as e:
                error(f"Watcher error, backing off: {e}")
                time.sleep(3)

File: dockerfiles/attestation-sidecar/app/internal/blockchain/event_watcher.go
package blockchain

import (
	"context"
	"encoding/json"
	"fmt"
	"math/big"
	"os"
	"path/filepath"
	"sort"
	"time"

	"app/internal/logger"

	"github.com/ethereum/go-ethereum"
	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/core/types"
)

// EventHandler is the callback signature for processing logs.
type EventHandler func(log types.Log, parsedData map[string]interface{}) error

type EventWatcher struct {
	client         *BlockchainClient
	eventName      string
	eventABI       abi.Event
	address        common.Address
	topics         [][]common.Hash
	checkpointPath string
	confirmations  uint64
	batchSize      uint64
	pollInterval   time.Duration

	// State
	fromBlock *big.Int
	seenKeys  map[string]bool
}

// CheckpointData matches the JSON structure of the Python version
type CheckpointData struct {
	FromBlock uint64   `json:"from_block"`
	SeenKeys  []string `json:"seen_keys"`
}

func NewEventWatcher(
	client *BlockchainClient,
	eventName string,
	checkpointPath string,
	confirmations uint64,
	batchSize uint64,
	pollIntervalSec float64,
) (*EventWatcher, error) {

	// 1. Find Event ABI
	evABI, ok := client.contractABI.Events[eventName]
	if !ok {
		return nil, fmt.Errorf("event '%s' not found in contract ABI", eventName)
	}

	// 2. Prepare Topics (Topic[0] is the event signature hash)
	topic0 := evABI.ID
	topics := [][]common.Hash{{topic0}}

	// 3. Init Watcher
	watcher := &EventWatcher{
		client:         client,
		eventName:      eventName,
		eventABI:       evABI,
		address:        client.contractAddress,
		topics:         topics,
		checkpointPath: checkpointPath,
		confirmations:  confirmations,
		batchSize:      batchSize,
		pollInterval:   time.Duration(pollIntervalSec * float64(time.Second)),
		seenKeys:       make(map[string]bool),
		fromBlock:      big.NewInt(0),
	}

	// 4. Load Checkpoint
	watcher.loadCheckpoint()

	return watcher, nil
}

func (ew *EventWatcher) loadCheckpoint() {
	if _, err := os.Stat(ew.checkpointPath); err == nil {
		fileBytes, err := os.ReadFile(ew.checkpointPath)
		if err == nil {
			var cp CheckpointData
			if err := json.Unmarshal(fileBytes, &cp); err == nil {
				ew.fromBlock = big.NewInt(int64(cp.FromBlock))
				for _, k := range cp.SeenKeys {
					ew.seenKeys[k] = true
				}
				logger.Debug("Loaded checkpoint for %s: from_block=%d, seen=%d", ew.eventName, cp.FromBlock, len(cp.SeenKeys))
				return
			}
		}
	}
	// Fallback logic
	header, err := ew.client.client.HeaderByNumber(context.Background(), nil)
	if err == nil && header != nil {
		ew.fromBlock = header.Number
	}
}

func (ew *EventWatcher) saveCheckpoint() {
	dir := filepath.Dir(ew.checkpointPath)
	if dir != "" {
		os.MkdirAll(dir, 0755)
	}

	keys := make([]string, 0, len(ew.seenKeys))
	for k := range ew.seenKeys {
		keys = append(keys, k)
	}

	cp := CheckpointData{
		FromBlock: ew.fromBlock.Uint64(),
		SeenKeys:  keys,
	}

	bytes, _ := json.MarshalIndent(cp, "", "  ")
	_ = os.WriteFile(ew.checkpointPath, bytes, 0644)
}

func (ew *EventWatcher) latestSafeBlock() *big.Int {
	header, err := ew.client.client.HeaderByNumber(context.Background(), nil)
	if err != nil {
		return big.NewInt(0)
	}
	current := header.Number.Uint64()
	if current < ew.confirmations {
		return big.NewInt(0)
	}
	return big.NewInt(int64(current - ew.confirmations))
}

func (ew *EventWatcher) uniqueKey(log types.Log) string {
	return fmt.Sprintf("%s-%d", log.BlockHash.Hex(), log.Index)
}

// Run starts the infinite polling loop. It blocks, so run in a goroutine.
func (ew *EventWatcher) Run(handler EventHandler) {
	logger.Info("EventWatcher start: event=%s address=%s from_block=%s confirm=%d",
		ew.eventName, ew.address.Hex(), ew.fromBlock.String(), ew.confirmations)

	for {
		safeTo := ew.latestSafeBlock()

		// If fromBlock > safeTo, we are ahead of safe chain, wait.
		if ew.fromBlock.Cmp(safeTo) > 0 {
			time.Sleep(ew.pollInterval)
			continue
		}

		// Loop to catch up range by range
		for ew.fromBlock.Cmp(safeTo) <= 0 {
			// toBlock = min(fromBlock + batchSize - 1, safeTo)
			batch := new(big.Int).SetUint64(ew.batchSize - 1)
			toBlock := new(big.Int).Add(ew.fromBlock, batch)
			if toBlock.Cmp(safeTo) > 0 {
				toBlock.Set(safeTo)
			}

			// Fetch Logs
			query := ethereum.FilterQuery{
				FromBlock: ew.fromBlock,
				ToBlock:   toBlock,
				Addresses: []common.Address{ew.address},
				Topics:    ew.topics,
			}

			logs, err := ew.client.client.FilterLogs(context.Background(), query)
			if err != nil {
				logger.Error("Watcher fetch error: %v. Backing off...", err)
				time.Sleep(3 * time.Second)
				break // break inner loop to retry
			}

			// Sort logs strictly by BlockNumber then Index
			sort.Slice(logs, func(i, j int) bool {
				if logs[i].BlockNumber == logs[j].BlockNumber {
					return logs[i].Index < logs[j].Index
				}
				return logs[i].BlockNumber < logs[j].BlockNumber
			})

			// Process
			for _, log := range logs {
				key := ew.uniqueKey(log)
				if ew.seenKeys[key] {
					continue
				}

				// --- Full Argument Decoding ---
				parsedData := make(map[string]interface{})

				// 1. Unpack non-indexed fields (Data)
				if len(log.Data) > 0 {
					if err := ew.eventABI.Inputs.UnpackIntoMap(parsedData, log.Data); err != nil {
						logger.Error("Failed to unpack non-indexed args for %s: %v", ew.eventName, err)
						continue
					}
				}

				// 2. Unpack indexed fields (Topics)
				// Topic[0] is signature. Args start at Topic[1].
				topicIndex := 1
				for _, arg := range ew.eventABI.Inputs {
					if arg.Indexed {
						if topicIndex >= len(log.Topics) {
							break
						}
						
						// Manual extraction based on type
						switch arg.Type.T {
						case abi.AddressTy:
							parsedData[arg.Name] = common.HexToAddress(log.Topics[topicIndex].Hex())
						case abi.IntTy, abi.UintTy:
							parsedData[arg.Name] = log.Topics[topicIndex].Big()
						case abi.BoolTy:
							parsedData[arg.Name] = (log.Topics[topicIndex].Big().Cmp(big.NewInt(0)) != 0)
						default:
							// Fallback: raw hash for strings/bytes/arrays
							parsedData[arg.Name] = log.Topics[topicIndex]
						}
						topicIndex++
					}
				}

				// Handle
				if err := handler(log, parsedData); err != nil {
					logger.Error("Handler error for log %s: %v", key, err)
				}
				ew.seenKeys[key] = true
			}

			// Clear seen keys and advance
			ew.seenKeys = make(map[string]bool)
			ew.fromBlock.Add(toBlock, big.NewInt(1))
			ew.saveCheckpoint()

			// Refresh safeTo
			safeTo = ew.latestSafeBlock()
		}

		time.Sleep(ew.pollInterval)
	}
}